File: src/main/java/io/github/svegon/capi/event/render/GUIRenderListener.java
package io.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.gui.DrawContext;

public interface GUIRenderListener {
    Event<GUIRenderListener> EVENT = EventFactory.createArrayBacked(GUIRenderListener.class,
            (startTime, tick, context, mouseX, mouseY, tickDelta) -> {},
            listeners -> (startTime, tick, context, mouseX, mouseY, tickDelta) -> {
                for (GUIRenderListener listener : listeners) {
                    listener.onGUIRender(startTime, tick, context, mouseX, mouseY, tickDelta);
                }
            });

    void onGUIRender(long startTime, boolean tick, DrawContext context, int mouseX, int mouseY, float tickDelta);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/GUIRenderListener.kt
package io.github.svegon.capi.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.gui.DrawContext
import java.util.function.Function

interface GUIRenderListener {
    fun onGUIRender(startTime: Long, tick: Boolean, context: DrawContext?, mouseX: Int, mouseY: Int, tickDelta: Float)

    companion object {
        val EVENT: Event<GUIRenderListener?> = EventFactory.createArrayBacked<GUIRenderListener?>(
            GUIRenderListener::class.java,
            GUIRenderListener { startTime: Long, tick: Boolean, context: DrawContext?, mouseX: Int, mouseY: Int, tickDelta: Float -> },
            Function<Array<GUIRenderListener?>, GUIRenderListener?> { listeners: Array<GUIRenderListener?>? ->
                GUIRenderListener { startTime: Long, tick: Boolean, context: DrawContext?, mouseX: Int, mouseY: Int, tickDelta: Float ->
                    for (listener in listeners) {
                        listener.onGUIRender(startTime, tick, context, mouseX, mouseY, tickDelta)
                    }
                }
            })
    }
}
